src/app/helpers/paginationTemplate/paginationTemplate.ts

```diff
@@ -0,0 +1,198 @@
+     * Default filter values, may be overridden by later requests
+     */
+    private defaultFilter: InnerFilter<I> = {}
+  ) {
+    super();
+  }
+
+  public ngOnInit() {
+    // Set pagination defaults
+    this.config.maxSize = 3;
+    this.config.pageSize = defaultApiPageSize;
+    this.config.rotate = true;
+    this.displayPagination = false;
+
+    this._page = 1;
+
+    this.apiRequest$
+      .pipe(
+        mergeMap(() => this.getModels()),
+        takeUntil(this.unsubscribe)
+      )
+      .subscribe(noop, (error: ApiErrorDetails) => {
+        console.error("Scroll Template Failure: ", error);
+        this.error = error;
+        this.loading = false;
+      });
+
+    this.updateFromUrl();
+
+    this.router.events
+      .pipe(
+        filter((event) => event instanceof NavigationEnd),
+        takeUntil(this.unsubscribe)
+      )
+      .subscribe(() => this.updateFromUrl(), noop);
+  }
+
+  /**
+   * Read the previous filter query and page number from the url,
+   * and update the page to match
+   */
+  private updateFromUrl() {
+    const params = this.route.snapshot.queryParams;
+    this.filter = params[queryKey] ?? "";
+    this._page = parseInt(params[pageKey], 10) || 1;
+    this.onFilter(this.filter, this._page);
+  }
+
+  public get page() {
+    return this._page;
+  }
+
+  /**
+   * Handle pagination events
+   */
+  public set page(page: number) {
+    this._page = page;
+    this.loading = true;
+    this.apiRequest$.next();
+    this.updateQueryParams(this._page, this.filter);
+  }
+
+  /**
+   * Handle filter events
+   */
+  public onFilter(input: string, page: number = 1) {
+    this._page = page;
+    this.filter = input;
+    this.loading = true;
+    this.apiRequest$.next();
+    this.updateQueryParams(this._page, input);
+  }
+
+  /**
+   * Update the url query parameters to contain the current page
+   * and filter query.
+   */
+  protected updateQueryParams(page: number, query?: string) {
+    const params = {};
+    if (page > 1) {
+      params[pageKey] = page;
+    }
+    if (query) {
+      params[queryKey] = query;
+    }
+
+    this.router.navigate([], {
+      relativeTo: this.route,
+      queryParams: params,
+    });
+  }
+
+  /**
+   * Retrieve the newest batch of models using the latest page and filter
+   * options. This will make a callback to the apiUpdate construction value
+   * once the models have been retrieved.
+   */
+  protected getModels(): Observable<void> {
+    const innerFilter = {
+      [this.filterKey]: { contains: this.filter },
+    } as { [P in keyof I]?: Subsets };
+
+    return this.api
+      .filter(
+        {
+          ...this.defaultFilter,
+          paging: { page: this._page },
+          filter: this.filter ? (innerFilter as any) : undefined,
+        },
+        ...this.apiParams()
+      )
+      .pipe(
+        map((models: M[]) => {
+          this.loading = false;
+          this.collectionSize = models?.[0]?.getMetadata()?.paging?.total || 0;
```

Comment:
probably better to use `null` to indicate an error value?

Reply:
This is not an error. It just means there are no models in the filter request, and i need to show that on the page